### PR TITLE
Shard python precommit

### PR DESCRIPTION
**Pull request by damccorm**
This splits the Python PreCommit into several suites. While the primary goal is debuggability, I used duration and test suites as a proxy for splitting things evenly, and this does have the tangential benefit of reducing our longest precommit to ~40 minutes (from ~80 minutes). This is the breakdown of the split, along with the number of test classes and execution time of each subfolder*:

```
- `runners` - 70 test classes, 6h 1m 52s
- `transforms` - 26 test classes, 4h 16m 56s
- `dataframe` - 11 test classes, 67m 55s
- `examples` - 68 test classes, 54m 38s

------------------------------------------------------------------------------------------------------------------------
All items below have been left in the main precommit. They add up to 73 test classes, 38m 184s of tests
------------------------------------------------------------------------------------------------------------------------
- `typehints` - 12 test classes, 23m 5s
- `./*.py` - 2 test classes, 19m 1s
- `testing` - 12 test classes, 8m 29s
- `ml` - 15 test classes, 5m 54s
- `utils` - 14 test classes, 1m 17s
- `coders` 9 test classes, 1m 10s
- `options` - 3 test classes, 39s
- `tools` - 1 test class, 25s
- `internal` - 7 test classes, 5s
- `portability` - 0 test classes, 0s
- `./*.py` - 2 test classes, 19m 1s
```

*Note that tests are run in parallel, so in reality its much faster than the time values above.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by damccorm** [comment]
Run Seed Job

**Comment by damccorm** [comment]
Run Python Internal Precommit

**Comment by damccorm** [comment]
Run Seed Job

**Comment by damccorm** [comment]
Run Python_Internal PreCommit

**Comment by damccorm** [comment]
Run Seed Job

**Comment by damccorm** [comment]
Run Python_Internal PreCommit

**Comment by damccorm** [comment]
Run Seed Job

**Comment by damccorm** [comment]
Run Seed Job

**Comment by damccorm** [comment]
Run Python PreCommit

**Comment by damccorm** [comment]
Run Python_Internal PreCommit

**Comment by damccorm** [comment]
Run Python Integration PreCommit

**Comment by damccorm** [comment]
Run Python Integration PreCommit

**Comment by damccorm** [comment]
Run Seed Job

**Comment by damccorm** [comment]
Run Python Integration PreCommit

**Comment by damccorm** [comment]
Run Python PreCommit

**Comment by damccorm** [comment]
Run Python Integration PreCommit

**Comment by damccorm** [comment]
Run Seed Job

**Comment by damccorm** [comment]
Run Python Integration PreCommit

**Comment by damccorm** [comment]
Run Python_Integration PreCommit

**Comment by damccorm** [comment]
Run Seed Job

**Comment by damccorm** [comment]
Run Python PreCommit

**Comment by damccorm** [comment]
Run Python_Integration PreCommit

**Comment by damccorm** [comment]
Run Python_Coders PreCommit

**Comment by damccorm** [comment]
Run Python_Dataframe PreCommit

**Comment by damccorm** [comment]
Run Python_Examples PreCommit

**Comment by damccorm** [comment]
Run Python_Internal PreCommit

**Comment by damccorm** [comment]
Run Seed Job

**Comment by damccorm** [comment]
Run Seed Job

**Comment by damccorm** [comment]
Run Python PreCommit

**Comment by damccorm** [comment]
Run Python_Integration PreCommit

**Comment by damccorm** [comment]
Run Python_Coders PreCommit

**Comment by damccorm** [comment]
Run Python_Examples PreCommit

**Comment by damccorm** [comment]
Run Python_Internal PreCommit

**Comment by damccorm** [comment]
Run Seed Job

**Comment by damccorm** [comment]
Run Python PreCommit

**Comment by damccorm** [comment]
Run Python_Integration PreCommit

**Comment by damccorm** [comment]
Run Python_Coders PreCommit

**Comment by damccorm** [comment]
Run Python_Dataframe PreCommit

**Comment by damccorm** [comment]
Run Python_Examples PreCommit

**Comment by damccorm** [comment]
Run Python_Internal PreCommit

**Comment by damccorm** [comment]
Run Python_Options PreCommit

**Comment by damccorm** [comment]
Run Python_Portability PreCommit

**Comment by damccorm** [comment]
Run Python_Runners PreCommit

**Comment by damccorm** [comment]
Run Python_Testing PreCommit

**Comment by damccorm** [comment]
Run Python_Tools PreCommit

**Comment by damccorm** [comment]
Run Python_Transforms PreCommit

**Comment by damccorm** [comment]
Run Python_Typehints PreCommit

**Comment by damccorm** [comment]
Run Python_Utils PreCommit

**Comment by damccorm** [comment]
Run Seed Job

**Comment by apilloud** [comment]
FYI, you've broken `Run PythonLint PreCommit` which now runs the following jobs:

Python_Integration ("Run Python_Integration PreCommit") — Build finished.
[Details](https://ci-beam.apache.org/job/beam_PreCommit_Python_Integration_Commit/8/)
@asf-ci
PythonLint ("Run PythonLint PreCommit") Pending — Build started for merge commit.
[Details](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Phrase/265/)
@asf-ci
Python_Dataframe ("Run Python_Dataframe PreCommit") Pending — Build started for merge commit.
[Details](https://ci-beam.apache.org/job/beam_PreCommit_Python_Dataframe_Commit/5/)
@asf-ci
Python_Examples ("Run Python_Examples PreCommit") Pending — Build started for merge commit.
[Details](https://ci-beam.apache.org/job/beam_PreCommit_Python_Examples_Commit/5/)
@asf-ci
Python_ML ("Run Python_ML PreCommit") Pending — Build started for merge commit.
[Details](https://ci-beam.apache.org/job/beam_PreCommit_Python_ML_Commit/1/)
@asf-ci
Python_Options ("Run Python_Options PreCommit") Pending — Build started for merge commit.
[Details](https://ci-beam.apache.org/job/beam_PreCommit_Python_Options_Commit/2/)
@asf-ci
Python_Portability ("Run Python_Portability PreCommit") Pending — Build started for merge commit.
[Details](https://ci-beam.apache.org/job/beam_PreCommit_Python_Portability_Commit/2/)
@asf-ci
Python_Runners ("Run Python_Runners PreCommit") Pending — Build started for merge commit.
[Details](https://ci-beam.apache.org/job/beam_PreCommit_Python_Runners_Commit/2/)
@asf-ci
Python_Testing ("Run Python_Testing PreCommit") Pending — Build started for merge commit.
[Details](https://ci-beam.apache.org/job/beam_PreCommit_Python_Testing_Commit/2/)
@asf-ci
Python_Tools ("Run Python_Tools PreCommit") Pending — Build started for merge commit.
[Details](https://ci-beam.apache.org/job/beam_PreCommit_Python_Tools_Commit/2/)
@asf-ci
Python_Transforms ("Run Python_Transforms PreCommit") Pending — Build started for merge commit.
[Details](https://ci-beam.apache.org/job/beam_PreCommit_Python_Transforms_Commit/2/)
@asf-ci
Python_Typehints ("Run Python_Typehints PreCommit") Pending — Build started for merge commit.
[Details](https://ci-beam.apache.org/job/beam_PreCommit_Python_Typehints_Commit/2/)
@asf-ci
Python_Utils ("Run Python_Utils PreCommit") Pending — Build started for merge commit.

**Comment by damccorm** [comment]
> FYI, you've broken `Run PythonLint PreCommit` which now runs the following jobs:
> 
> Python_Integration ("Run Python_Integration PreCommit") — Build finished. [Details](https://ci-beam.apache.org/job/beam_PreCommit_Python_Integration_Commit/8/) @asf-ci PythonLint ("Run PythonLint PreCommit") Pending — Build started for merge commit. [Details](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Phrase/265/) @asf-ci Python_Dataframe ("Run Python_Dataframe PreCommit") Pending — Build started for merge commit. [Details](https://ci-beam.apache.org/job/beam_PreCommit_Python_Dataframe_Commit/5/) @asf-ci Python_Examples ("Run Python_Examples PreCommit") Pending — Build started for merge commit. [Details](https://ci-beam.apache.org/job/beam_PreCommit_Python_Examples_Commit/5/) @asf-ci Python_ML ("Run Python_ML PreCommit") Pending — Build started for merge commit. [Details](https://ci-beam.apache.org/job/beam_PreCommit_Python_ML_Commit/1/) @asf-ci Python_Options ("Run Python_Options PreCommit") Pending — Build started for merge commit. [Details](https://ci-beam.apache.org/job/beam_PreCommit_Python_Options_Commit/2/) @asf-ci Python_Portability ("Run Python_Portability PreCommit") Pending — Build started for merge commit. [Details](https://ci-beam.apache.org/job/beam_PreCommit_Python_Portability_Commit/2/) @asf-ci Python_Runners ("Run Python_Runners PreCommit") Pending — Build started for merge commit. [Details](https://ci-beam.apache.org/job/beam_PreCommit_Python_Runners_Commit/2/) @asf-ci Python_Testing ("Run Python_Testing PreCommit") Pending — Build started for merge commit. [Details](https://ci-beam.apache.org/job/beam_PreCommit_Python_Testing_Commit/2/) @asf-ci Python_Tools ("Run Python_Tools PreCommit") Pending — Build started for merge commit. [Details](https://ci-beam.apache.org/job/beam_PreCommit_Python_Tools_Commit/2/) @asf-ci Python_Transforms ("Run Python_Transforms PreCommit") Pending — Build started for merge commit. [Details](https://ci-beam.apache.org/job/beam_PreCommit_Python_Transforms_Commit/2/) @asf-ci Python_Typehints ("Run Python_Typehints PreCommit") Pending — Build started for merge commit. [Details](https://ci-beam.apache.org/job/beam_PreCommit_Python_Typehints_Commit/2/) @asf-ci Python_Utils ("Run Python_Utils PreCommit") Pending — Build started for merge commit.

I think this is probably just a consequence of the new triggers causing these to run? I'll run a seed job off of master anyways though

**Comment by damccorm** [comment]
Run Python PreCommit

**Comment by damccorm** [comment]
Run Seed Job

**Comment by damccorm** [comment]
Run Python PreCommit

**Comment by damccorm** [comment]
Run Python_Integration PreCommit

**Comment by damccorm** [comment]
Run Python_Coders PreCommit

**Comment by damccorm** [comment]
Run Python_Dataframe PreCommit

**Comment by damccorm** [comment]
Run Python_Examples PreCommit

**Comment by damccorm** [comment]
Run Python_Internal PreCommit

**Comment by damccorm** [comment]
Run Python_ML PreCommit

**Comment by damccorm** [comment]
Run Python_Options PreCommit

**Comment by damccorm** [comment]
Run Python_Portability PreCommit

**Comment by damccorm** [comment]
Run Python_Runners PreCommit

**Comment by damccorm** [comment]
Run Python_Testing PreCommit

**Comment by damccorm** [comment]
Run Python_Tools PreCommit

**Comment by damccorm** [comment]
Run Python_Transforms PreCommit

**Comment by damccorm** [comment]
Run Python_Typehints PreCommit

**Comment by damccorm** [comment]
Run Python_Utils PreCommit

**Comment by damccorm** [comment]
Run Seed Job

**Comment by Abacn** [comment]
Hi, noticed this when jobs are triggered in other PR. To my experience the test of a single Python version does not take long, but the current PreCommit test suite is running a full version matrix (Py37, Py38, Py39, Py310), in addition to a full version matrix of pyarrow, pytorch, pandas.

Given that we are going to migrate to GitHub Actions, I would suggest least effort split as splitting to each precommit of each python version (Py37 to Py310), in addition to another precommit job running a full version matrix of packages pyarrow, pytorch, pandas. Because GitHub Action would also split a job by py version through matrix.

**Comment by damccorm** [comment]
@Abacn I don't think that approach fully accomplishes what I'm aiming for here though. There are several advantages of sharding tests:

[1] Faster execution time on individual Precommits so that we get a faster fail/succeed signal.
[2] Lower rate of flakes per suite (makes retrying easier).
[3] Better signal on which test suites are consistently flaky.
[4] Gating precommits to tighter directories (e.g. don't run dataframes tests if only IOs have changed)

I'm mostly going after item number [3] in this PR; splitting by language version doesn't really help with [3] since test suites generally fail together. This split would help with debuggability by providing a tighter signal, and hopefully shrinking our test surface enough that the test analyzer plugin works. It also comes with some tangential benefits to [1] and [2], but that isn't really the goal.

Splitting by language version and/or tool version will help with items [1] and [2], but there's not a very clean way to logically represent that kind of matrix as Jenkins jobs so I'd probably prefer to defer that until after the GHA migration.

Its also worth noting that this isn't an accurate final view of this PR and I'm planning on consolidating back down to several (4ish) jobs.

**Comment by damccorm** [comment]
Run Seed Job

**Comment by Abacn** [comment]
Yeah I get (2,3), strongly agree this approach is advantage in getting better signal. Did not consider this premise. Thanks!

**Comment by damccorm** [comment]
Run Python PreCommit

**Comment by damccorm** [comment]
Run Python_Runners PreCommit

**Comment by damccorm** [comment]
Run Python_Transforms PreCommit

**Comment by damccorm** [comment]
Run Python_Dataframe PreCommit

**Comment by damccorm** [comment]
Run Python_Examples PreCommit

**Comment by damccorm** [comment]
Run Python_Integration PreCommit

**Comment by damccorm** [comment]
Run Seed Job

**Comment by damccorm** [comment]
Run Python PreCommit

**Comment by damccorm** [comment]
Run Python_Runners PreCommit

**Comment by damccorm** [comment]
Run Python_Transforms PreCommit

**Comment by damccorm** [comment]
Run Python_Dataframe PreCommit

**Comment by damccorm** [comment]
Run Python_Examples PreCommit

**Comment by damccorm** [comment]
Run Python_Integration PreCommit

**Comment by damccorm** [comment]
Run Python_Runners PreCommit

**Comment by damccorm** [comment]
Run Seed Job

**Comment by damccorm** [comment]
Run Python_Runners PreCommit

**Comment by damccorm** [comment]
Run Seed Job

**Comment by damccorm** [comment]
Run Python_Runners PreCommit

**Comment by damccorm** [comment]
R: @Abacn

File: sdks/python/test-suites/tox/py38/build.gradle [hunk 1]
```diff
@@ -26,6 +26,10 @@ applyPythonNature()
 // Required to setup a Python 3 virtualenv and task names.
 pythonVersion = '3.8'
 
+def posargs = project.findProperty("posargs") ?: ""
+def runDataframesTests = (posargs == "" || posargs.contains("dataframes"))
```
Comment (Abacn) [inline sdks/python/test-suites/tox/py38/build.gradle:30]: py38/build.gradle has more content than other versions and these additional contents are for precommit running on a single version. We should gather these single version lines together (under comment TODO(BEAM-12000) below)  to make future version bump easier.

**Review by Abacn — COMMENTED** [review]
Thanks! The large python precommit has long been painful. Curious about how long does it take now for job_PreCommit_Python if this is in? I see the check still runs on full test.

File: .test-infra/jenkins/job_PreCommit_Python.groovy [hunk 2]
```diff
@@ -22,6 +22,9 @@ PrecommitJobBuilder builder = new PrecommitJobBuilder(
     scope: this,
     nameBase: 'Python',
     gradleTask: ':pythonPreCommit',
+    gradleSwitches: [
```
[the remark this answers is not in the document]
Comment (Abacn) [reply .test-infra/jenkins/job_PreCommit_Python.groovy:25]: Also tested on my draft pr (https://ci-beam.apache.org/job/beam_PreCommit_Python_Commit/26409/) and found understand the obstacle of using `--ignore`. Although run_pytest.sh has pytest_args in addition to posargs, tox only has posargs that is passed from gradle script. Currently, feed `--ignore` in posargs will get this error:

-----

ERROR: module or package not found: '--ignore=apache_beam/coders (missing __init__.py?)

ERROR: InvocationError for command /home/jenkins/jenkins-slave/workspace/beam_PreCommit_Python_Commit/src/sdks/python/test-suites/tox/py37/build/srcs/sdks/python/scripts/run_pytest.sh py37-cython ''"'"'--ignore=apache_beam/coders --ignore=apache_beam/dataframe --ignore=apache_beam/examples --ignore=apache_beam/internal --ignore=apache_beam/io --ignore=apache_beam/metrics --ignore=apache_beam/ml'"'"'' (exited with code 4)

-----

**Review by Abacn — APPROVED** [review]
This LGTM. Let's see the effects of this change. 

`ModuleNotFoundError: No module named 'typing_extensions'` in pythonlint and python precommit are irrelevant, though these have been permared since today. Likely some transitive dependency change caused this.

**Comment by damccorm** [comment]
Run PythonLint PreCommit

**Comment by damccorm** [comment]
Run Python PreCommit

File: .test-infra/jenkins/job_PreCommit_Python.groovy (hunk 2 above)
[the remark this answers is not in the document]
Comment (damccorm) [reply .test-infra/jenkins/job_PreCommit_Python.groovy:25]: Filed #24813

**Comment by Abacn** [comment]
Looks like this PR breaks Python PreCommit, https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/6546/
```
ERROR: module or package not found: 'apache_beam/*.py (missing __init__.py?)

ERROR: InvocationError for command /home/jenkins/jenkins-slave/workspace/beam_PreCommit_Python_Cron/src/sdks/python/test-suites/tox/py39/build/srcs/sdks/python/scripts/run_pytest.sh py39-cython ''"'"'apache_beam/*.py apache_beam/coders apache_beam/internal apache_beam/io apache_beam/metrics apache_beam/ml apache_beam/options apache_beam/portability apache_beam/testing apache_beam/tools apache_beam/typehints apache_beam/utils'"'"'' (exited with code 4)
```
the error is similar with what I saw in https://github.com/apache/beam/pull/24204#discussion_r1058662550

**Pull request merged**